src/function/built_in_functions.cpp

```diff
@@ -119,55 +119,64 @@ AggregateFunction* BuiltInFunctions::matchAggregateFunction(
-            return 0;
-        case LogicalTypeID::RDF_VARIANT:
-            // RDF_VARIANT can be cast to any type.
-            return 1;
-        case LogicalTypeID::INT64:
-            return castInt64(targetTypeID);
-        case LogicalTypeID::INT32:
-            return castInt32(targetTypeID);
-        case LogicalTypeID::INT16:
-            return castInt16(targetTypeID);
-        case LogicalTypeID::INT8:
-            return castInt8(targetTypeID);
-        case LogicalTypeID::UINT64:
-            return castUInt64(targetTypeID);
-        case LogicalTypeID::UINT32:
-            return castUInt32(targetTypeID);
-        case LogicalTypeID::UINT16:
-            return castUInt16(targetTypeID);
-        case LogicalTypeID::UINT8:
-            return castUInt8(targetTypeID);
-        case LogicalTypeID::INT128:
-            return castInt128(targetTypeID);
-        case LogicalTypeID::DOUBLE:
-            return castDouble(targetTypeID);
-        case LogicalTypeID::FLOAT:
-            return castFloat(targetTypeID);
-        case LogicalTypeID::DATE:
-            return castDate(targetTypeID);
-        case LogicalTypeID::SERIAL:
-            return castSerial(targetTypeID);
-        case LogicalTypeID::TIMESTAMP_SEC:
-        case LogicalTypeID::TIMESTAMP_MS:
-        case LogicalTypeID::TIMESTAMP_NS:
-        case LogicalTypeID::TIMESTAMP_TZ:
-            if (targetTypeID == LogicalTypeID::TIMESTAMP) {
-                return 101;
-            }
+        // Any other inputTypeID can be cast to String, but this cast has a high cost
+        return getTargetTypeCost(targetTypeID);
+    }
+    // for numeric, date, timestamp, SERIAL
+    switch (inputTypeID) {
+    case LogicalTypeID::INT64:
+        return castInt64(targetTypeID);
+    case LogicalTypeID::INT32:
+        return castInt32(targetTypeID);
+    case LogicalTypeID::INT16:
+        return castInt16(targetTypeID);
+    case LogicalTypeID::INT8:
+        return castInt8(targetTypeID);
+    case LogicalTypeID::UINT64:
+        return castUInt64(targetTypeID);
+    case LogicalTypeID::UINT32:
+        return castUInt32(targetTypeID);
+    case LogicalTypeID::UINT16:
+        return castUInt16(targetTypeID);
+    case LogicalTypeID::UINT8:
+        return castUInt8(targetTypeID);
+    case LogicalTypeID::INT128:
+        return castInt128(targetTypeID);
+    case LogicalTypeID::DOUBLE:
+        return castDouble(targetTypeID);
+    case LogicalTypeID::FLOAT:
+        return castFloat(targetTypeID);
+    case LogicalTypeID::DATE:
+        return castDate(targetTypeID);
+    case LogicalTypeID::SERIAL:
+        return castSerial(targetTypeID);
+    case LogicalTypeID::TIMESTAMP_SEC:
+    case LogicalTypeID::TIMESTAMP_MS:
+    case LogicalTypeID::TIMESTAMP_NS:
+    case LogicalTypeID::TIMESTAMP_TZ:
         // currently don't allow timestamp to other timestamp types
-        default:
-            return UNDEFINED_CAST_COST;
+        if (targetTypeID == LogicalTypeID::TIMESTAMP) {
```

Comment:
ditto we should do castTimestamp for consistency

Reply:
leave this to the next pr